File: app/src/main/cpp/ime/nlp/suggestion_list.cpp
/*
 * Copyright (C) 2021 Patrick Goldinger
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "suggestion_list.h"

#include <utility>

using namespace ime::nlp;

SuggestionList::SuggestionList(size_t _maxSize) :
    maxSize(_maxSize), internalSize(0), tokens(_maxSize), isPrimaryTokenAutoInsert(false)
{ }

SuggestionList::~SuggestionList() = default;

bool SuggestionList::add(word_t &&word, freq_t &&freq) {
    auto entryIndex = indexOfWord(word);
    if (entryIndex.has_value()) {
        // Word exists already
        auto entry = tokens[entryIndex.value()];
        if (entry.freq < freq) {
            // Need to update freq
            entry.freq = freq;
        } else {
            return false;
        }
    } else {
        if (internalSize < maxSize) {
            tokens[internalSize++] = WeightedToken(std::move(word), freq);
        } else {
            auto last = tokens[internalSize - 1];
            if (last.freq < freq) {
                tokens[internalSize - 1] = WeightedToken(std::move(word), freq);
            } else {
                return false;
            }
        }
    }
    std::sort(tokens.begin(), tokens.begin() + internalSize, std::greater<>());
    return true;
}

void SuggestionList::clear() {
    internalSize = 0;
    isPrimaryTokenAutoInsert = false;
}

bool SuggestionList::contains(const WeightedToken &element) const {
    return indexOf(element).has_value();
}

bool SuggestionList::containsWord(const word_t &word) const {
    return indexOfWord(word).has_value();
}

const WeightedToken *SuggestionList::get(size_t index) const {
    if (index < 0 || index >= internalSize) return nullptr;
    return &tokens[index];
}

std::optional<size_t> SuggestionList::indexOf(const WeightedToken &element) const {
    for (size_t n = 0; n < internalSize; n++) {
        if (element == tokens[n]) {
            return n;
        }
    }
    return std::nullopt;
}

std::optional<size_t> SuggestionList::indexOfWord(const word_t &word) const {
    for (size_t n = 0; n < internalSize; n++) {
        if (word == tokens[n].data) {
            return n;
        }
    }
    return std::nullopt;
}

bool SuggestionList::isEmpty() const {
    return internalSize == 0;
}

size_t SuggestionList::size() const {
    return internalSize;
}

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/nlp/SuggestionList.kt
/*
 * Copyright (C) 2021 Patrick Goldinger
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.patrickgold.florisboard.ime.nlp

import dev.patrickgold.florisboard.lib.NativeInstanceWrapper
import dev.patrickgold.florisboard.lib.NativePtr

@JvmInline
value class SuggestionList private constructor(
    private val _nativePtr: NativePtr
) : Collection<String>, NativeInstanceWrapper {
    companion object {
        fun new(maxSize: Int): SuggestionList {
            val nativePtr = nativeInitialize(maxSize)
            return SuggestionList(nativePtr)
        }

        external fun nativeInitialize(maxSize: Int): NativePtr
        external fun nativeDispose(nativePtr: NativePtr)

        external fun nativeAdd(nativePtr: NativePtr, word: Word, freq: Freq): Boolean
        external fun nativeClear(nativePtr: NativePtr)
        external fun nativeContains(nativePtr: NativePtr, element: Word): Boolean
        external fun nativeGetOrNull(nativePtr: NativePtr, index: Int): Word?
        external fun nativeGetIsPrimaryTokenAutoInsert(nativePtr: NativePtr): Boolean
        external fun nativeSetIsPrimaryTokenAutoInsert(nativePtr: NativePtr, v: Boolean)
        external fun nativeSize(nativePtr: NativePtr): Int
    }

    override val size: Int
        get() = nativeSize(_nativePtr)

    fun add(word: Word, freq: Freq): Boolean {
        return nativeAdd(_nativePtr, word, freq)
    }

    fun clear() {
        nativeClear(_nativePtr)
    }

    override fun contains(element: Word): Boolean {
        return nativeContains(_nativePtr, element)
    }

    override fun containsAll(elements: Collection<Word>): Boolean {
        elements.forEach { if (!contains(it)) return false }
        return true
    }

    @Throws(IndexOutOfBoundsException::class)
    operator fun get(index: Int): Word {
        val element = getOrNull(index)
        if (element == null) {
            throw IndexOutOfBoundsException("The specified index $index is not within the bounds of this list!")
        } else {
            return element
        }
    }

    fun getOrNull(index: Int): Word? {
        return nativeGetOrNull(_nativePtr, index)
    }

    override fun isEmpty(): Boolean = size <= 0

    val isPrimaryTokenAutoInsert: Boolean
        get() = nativeGetIsPrimaryTokenAutoInsert(_nativePtr)

    override fun iterator(): Iterator<Word> {
        return SuggestionListIterator(this)
    }

    override fun nativePtr(): NativePtr {
        return _nativePtr
    }

    override fun dispose() {
        nativeDispose(_nativePtr)
    }

    class SuggestionListIterator internal constructor (
        private val suggestionList: SuggestionList
    ) : Iterator<Word> {
        var index = 0

        override fun next(): Word = suggestionList[index++]

        override fun hasNext(): Boolean = suggestionList.getOrNull(index) != null
    }
}
